### DISCUSSION TITLE
Pc Connection

### DISCUSSION THREAD

**Discussion by ftsaci** [micbay/RaceTimerAndController#23] [General]
hello. it's me again :)  is it possible to print the result data to the serial monitor? so that the data can be collected via excel. my other question is, can the start stop commands be given via pc ?

**Comment by micbay** [comment]
Anything is possible, but the Arduino has limited available memory and this project is at the limit, which means trade-offs in existing functions must be made in order to implement new ones that require an additional library or more than a few extra lines of code.

That said, there are some things you can do to regain memory from the current v2.0 code.
- Reduce the saved lap count (memory used by saved lap times are directly related to how many lanes are enabled (set by `LANE_COUNT`), and how many laps are saved for each lane (set by `DEFAULT_MAX_STORED_LAPS`)
- Remove/comment out, the code for the Adafruit LED bar, if not using as a start light.
    - Comment out Adafruit code from line 47-73
    - Search for usage of variable `clearStartLight` and comment those lines out, or remove from any conditional statements used. NOTE: some places might be using this boolean flag with other variables, so don't just comment out the whole line of code if it is being used this way, just remove the `clearStartLight` variable from the expression.
    - Search for calls to the function `setBargraph()` and comment them out.

## Get Familiar With Using Serial on Arduino
I know you've said you are new to programming, this Arduino forum post can help you get familiar with Arduino serial port usage [Serial Input Basics - updated](https://forum.arduino.cc/t/serial-input-basics-updated/382007).

Also, I'd recommend that you first start with using a very basic serial sketch that you can find probably in the built in examples in the Arduino IDE, or on the web, and make sure you can connect, read, and write to the Arduino before moving on with trying to implement changes to the race timer sketch.

## Initialize Serial Connection
To initialize the serial connection, in the `setup()` function, call `Serial.begin(baudrate)`. This code already exists and can be uncommented on lines 1848-1849.

At this point you have serial communication and can send data to a computer with an open serial monitor that is using the same baudrate the connection was initialized with, set by using:
`Serial.println("whatever text in quotes, or number or string variable you want to print to the computer")`

## Printing Race Results
From here probably the best place to export results data would be to add a new case for an unused, keypad button (line 2209) in the `ResultsMenu` menu state (line 2182) that will execute a loop printing the data you want.

Instead of coding that loop all inside the menu state, it's probably better to create a function that does that for you and simply call that function when the chosen 'print to serial' key is pressed. Refer to the current `PrintResultsList()` function on line 1439, for some reference. You can modify that function to work with serial or make a new function, replacing the `PrintNumbers()`, and `PrintClock()` function calls, with versions that print the values out to serial instead of the display. Note that the current `PrintResultsList()` only has 3 lines available for display so it will not print a full list as is, you will need to change the `for` loop to print the full `listSize`.

This should give you some starting points to work out dumping data to a serial monitor, which will need to be manually copied or have some kind of data export function in the serial monitor your using to then save that data to a txt, excel, or other filetype.

## Send Command to Arduino from CPU
Regarding talking back to the device, in order to use a computer to operate a race, you can try to add in a `SerialRead()` inside the `Staging` state (line 2266) to react to receiving the command from your serial port. I'd recommend using a single character as that is how serial works, character by character. There might be built in support for full strings, depending on the tools you're using to monitor serial on the computer, but be aware that if the serial read on the Arduino side is not setup appropriately for receiving multiple values before executing on the input you may not be getting the behavior you want. So try a single letter before getting more complicated.

Starting at line 2302 where the keypad is polled for a `#` or `*` you can poll for the serial command in a similar manner and modify the conditional `if - else` statement on 2304 to also react to the serial command value.
To get the value from serial try something like the following, where an `s` sent from computer, means start and `e` means exit:
```
    boolean startRace = false;
    boolean exitToMenu = false;
    if (Serial.available() > 0) {
        char serialCmd = Serial.read();
        if (serialCmd == 's') startRace = true;
        if (serialCmd == 'e') exitToMenu = true;
    }
```
Integrated into the current code at line 2302 might look something like this:
```
      // After entry, monitor for user input to decide next state
      char key = keypad.getKey();
      boolean startRace = false;
      boolean exitToMenu = false;
      if (Serial.available() > 0) {
          char serialCmd = Serial.read();
          if (serialCmd == 's') startRace = true;
          if (serialCmd == 'e') exitToMenu = true;
      }
      // If Start button or '#' key is pressed, initiate a race, and switch to PreStart state.
      if( buttonPressed(startButtonPin) || key == '#' || startRace) {
        ChangeStateTo(PreStart);
        preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
        newRace = true;
        // Serial.println("dgbt");
      } else
      // If Pause button, or '*' key is pressed, exit back to the Main Menu.
      if( buttonPressed(pauseStopPin)  || key == '*' || exitToMenu) {
        // clear start light displays
        lc.clearDisplay(laneCount);
        setBargraph(LED_OFF);
        // Move to Menu state
        ChangeStateTo(Menu);
        currentMenu = MainMenu;
        // Serial.println("psbt");
      }
```

**Reply by ftsaci** [reply]
Thank you very much for your interest. To get information from the serial monitor, I added the following to line 1466

```
int leaderBoard[laneCount][2] = {};
unsigned long overallFastestTime = 99999999;
byte overallFastestRacer = 5;

// Liderlik tablosunu Serial Monitör'e yazdırma fonksiyonu
void PrintLeaderBoardToSerial() {
  Serial.println("Liderlik Tablosu:");

  for (byte i = 0; i < laneCount; i++) {
    Serial.print("Sıra ");
    Serial.print(i + 1);
    Serial.print(" | Yarışçı: ");
    Serial.print(Racers[leaderBoard[i][1]]);  // Yarışçı isimlerini yazdır

    Serial.print(" | Toplam Tur: ");
    Serial.print(leaderBoard[i][0] - 1);  // Toplam tur sayısını bir eksik göster

    // Eğer kayıtlı lap sürelerini de yazdırmak istiyorsanız
    Serial.print(" | Süreler: ");
    for (byte j = 0; j < lapCount[leaderBoard[i][1]]; j++) {
      Serial.print(fastestTimes[leaderBoard[i][1]][j] / 1000.0, 3);  // Süreleri saniye cinsinden yazdırma
      if (j < lapCount[leaderBoard[i][1]] - 1) {
        Serial.print(", ");
      }
    }
    Serial.println();
  }
}

// UpdateResultsMenu fonksiyonu
void UpdateResultsMenu(bool full = true) {
  switch (resultsMenuIdx) {
    case 0:
      {
        lcd.clear();
        PrintText(RESULTS_TOP_LBL, lcdDisp, 14, 15, false, 0);
        lcd.setCursor(17, 0);
        lcd.write('A');
        lcd.write(0);
        lcd.write('B');
        PrintResultsList(fastestTimes[0], fastestLaps[0], DEFAULT_MAX_STORED_LAPS);

        // Liderlik tablosunu Serial Monitör'e her yarış bitiminde yazdır
        PrintLeaderBoardToSerial();
      }
      break;
    case 1 ... laneCount:
      {
        if (laneEnableStatus[resultsMenuIdx] > 0) {
          if (titleA) {
            PrintText(RESULTS_RACER_LBL, lcdDisp, 16, 17, false, 0);
            PrintNumbers(resultsMenuIdx, 1, RESULTS_RACER_NUM_POS, lcdDisp, false, 0);
            PrintText(RESULTS_TOTAL_LBLA, lcdDisp, 19, 6, true, 2, false);
            PrintClock(racersTotalTime[resultsMenuIdx], 19, 6, 1, lcdDisp, 3, false);
          } else {
            PrintText(Racers[laneRacer[resultsMenuIdx]], lcdDisp, 16, 14, false, 0);
            PrintText(RESULTS_TOTAL_LBLB, lcdDisp, 19, 6, true, 2, false);
            PrintNumbers(((lapCount[resultsMenuIdx] == 0) ? 0 : (lapCount[resultsMenuIdx] - 1)), 6, 19, lcdDisp, false, 3);
          }
          if (full) {
            lcd.setCursor(17, 0);
            lcd.write('A');
            lcd.write(0);
            lcd.write('B');
            PrintResultsList(
              fastestTimes[resultsMenuIdx],
              fastestLaps[resultsMenuIdx],
              lapCount[resultsMenuIdx] < DEFAULT_MAX_STORED_LAPS ? lapCount[resultsMenuIdx] : DEFAULT_MAX_STORED_LAPS);
            PrintText(RESULTS_TOTAL_LBL, lcdDisp, 19, 6, true, 1, false);
          }
        } else {
          resultsMenuIdx++;
          UpdateResultsMenu(full);
        }
      }
      break;
    case laneCount + 1:
      {
        lcd.clear();
        PrintText(RESULTS_FINISH_LBL, lcdDisp, 19, 20, false, 0);
        PrintLeaderBoard(true);
      }
    default:
      break;
  }
}
```

. Now I can get end-of-race data from my PC without any problems. (I would also like to thank chatgpt for this :)) I wanted to add a function to line 1988 for remote start, as I tried before, but I could not succeed. The code you sent about this worked, but only on the race waiting page. The result I am trying to achieve is to be able to receive commands from the computer in the main menu.

**Comment by micbay** [comment]
I'm a bit confused on the state of your build and what specifically, if anything, you're still trying to implement. I gather you got the data output you want, but haven't quite gotten it to take a command from the screen you want, is that a correct assessment?

You should be able to take this general code and put it in the state you want to take cpu commands from
```
    // initiate a variable to hold a character read from serial port, with a default dummy value 'x'
    char serialCmd = 'x';
    // Make sure there is data in the serial buffer before trying to read from it
    if (Serial.available() > 0) {
        // read character into variable
        serialCmd = Serial.read();
    }

    // use a 'switch' or 'if' statement, or existing conditionals to check if 'serialCmd' matches a your command then do something
```
However, if you accidentally skip a part that initializes the race state, like we did with the initial implementation of the other discussion, that caused the 1st lap time error, it may not work as expected. If you look at the `StartRaceMenu` state (line 2139), then in the conditional that responds to a race start key, we set a race type and race initial variables. You are probably skipping that and need to move race setup code to where you actually initiate a race state.
```
           // From 'StartRaceMenu' 
            case 'A': case 'B': case 'C': {
              endLap = raceLaps;
              countingDown = false;
              if (key == 'A') raceType = Standard;
              else if (key == 'B') {raceType = Timed; countingDown = true;}
              else if (key == 'C') {raceType = Drag; endLap = 1;}
              ChangeStateTo(Staging);
            }
```

You could move the above into the main menu state and change the case values to be different serial cmd chars that can be used to initiate different race types. Maybe something like:
```
    // initiate a variable to hold a character read from serial port, use a dummy value 'x'
    char serialCmd = 'x';
    // Make sure there is data in the serial buffer before trying to read from it
    if (Serial.available() > 0) {
        // read character into variable
        serialCmd = Serial.read();
    }

    // use a 'switch' or 'if' statement, or existing conditionals to check if 'serialCmd' matches a your command then do something
    // An 'A' from cpu will start a standard race, 'B' will start a timed race, and 'C' will start a drag race
    switch (serialCmd) {
      case 'A': case 'B': case 'C': {
        endLap = raceLaps;
        countingDown = false;
        if (key == 'A') raceType = Standard;
        else if (key == 'B') {raceType = Timed; countingDown = true;}
        else if (key == 'C') {raceType = Drag; endLap = 1;}
        
        ChangeStateTo(PreStart);
        preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
        newRace = true;
      }
    }
    break;
    default:
    break;
```

Try adding this code above in the main menu state at line 1988. It's fine that we check for 'A', 'B', and 'C' for the keypad and for the CPU to do different things because the switch statements will be checking different variables.

**Comment by ftsaci** [comment]
yes my friend. The problem was as you first mentioned. I couldn't receive commands from the computer. I made an addition to the place you mentioned, as in your example. (Thanks again to chatgpt :) ) Sending commands from the computer is now working without any problems. Thank you very much. for everything.

```
switch (currentMenu) {
  // The 'MainMenu' is the default and top level menu in the menu tree.
  // Within each menu case is another switch for each available key input.
  case MainMenu:
    {
      if (entryFlag) {
        UpdateLCDMenu(MainText);
        // Update Racer LED Displays
        UpdateAllNamesOnLEDs();
        entryFlag = false;
      }

      // initiate a variable to hold a character read from serial port, use a dummy value 'x'
      char serialCmd = 'x';
      // Make sure there is data in the serial buffer before trying to read from it
      if (Serial.available() > 0) {
          // read character into variable
          serialCmd = Serial.read();
      }

      // Depending on selected option, change menu state accordingly and set entry flag.
      switch (key) {
        case 'A':
          currentMenu = SelectRacersMenu;
          entryFlag = true;
          break;
        case 'B':
          currentMenu = SettingsMenu;
          entryFlag = true;
          break;
        case 'C':
          currentMenu = StartRaceMenu;
          entryFlag = true;
          break;
        case 'D':
          currentMenu = ResultsMenu;
          resultsMenuIdx = 0;
          entryFlag = true;
          break;
        case '#':                
          endLap = raceLaps;
          countingDown = false;
          raceType = Standard;
          ChangeStateTo(PreStart);
          preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
          newRace = true;
          break;

        default:
          break;
      }  // END key switch

      // Check for Serial commands to start races
      switch (serialCmd) {
        case 'S':  // Start Standard Race
          raceType = Standard;
          countingDown = false;
          endLap = raceLaps;
          ChangeStateTo(PreStart);
          preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
          newRace = true;
          Serial.println("Yarış Başlatıldı");  // "Yarış Başlatıldı" mesajını Serial Monitör'e yazdır
          break;
        case 'B':  // Start Timed Race
          raceType = Timed;
          countingDown = true;
          ChangeStateTo(PreStart);
          preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
          newRace = true;
          break;
        case 'C':  // Start Drag Race
          raceType = Drag;
          endLap = 1;
          countingDown = false;
          ChangeStateTo(PreStart);
          preStartTimerDrag = (curMillis % DRAG_PRESTART_RNDM) + DRAG_PRESTART_CNTDWN_BASE;
          newRace = true;
          break;
        default:
          break;
      }  // END Serial command switch
```
